### PR TITLE
fix(opencode): complete API key connect when the runtime probe cannot verify it

### PR DESCRIPTION
**Pull request by NeatCode-Labs**
*(two commits; the second adds the setup-form connect path)*

## Problem

Pasting a valid Anthropic API key into the OpenCode provider connect dialog
fails with:

```
OpenCode could not verify provider anthropic with 3 model candidates:
anthropic/claude-sonnet-4-5: Not Found; anthropic/claude-haiku-4-5: Not Found;
anthropic/claude-opus-4-1: Not Found
```

The key is correct. It works everywhere else. It simply can never be connected
through this dialog.

Root cause is an ordering problem in the orchestrator's `connect-api-key`
command: it verifies a submitted key by probing catalog model candidates, but
OpenCode only activates a provider **after** its credential is in the auth
store. For store-activated providers the probe therefore runs in a scope where
the submitted key is not active yet, every candidate answers `Not Found`, the
command reports `auth-failed`, and the credential is never committed. The
verification can never pass, so the connect can never succeed.

The same failure reaches the provider **setup form**, which connects through
`runtime providers connect` with an `api` method rather than `connect-api-key`.
It hit the identical probe failure and surfaced the raw `Not Found` candidate
list to the user.

## Fix

When — and only when — a connect response carries that specific probe-failure
signature, and the provider is one the app can verify directly, the app does the
verification itself and completes the connect:

1. **Classify the failure.** `openCodeConnectVerifyFailure.ts` (new) matches the
   "could not verify provider" signature across the error message and every
   diagnostics field, after stripping terminal formatting and collapsing
   whitespace. The pattern is deliberately loose about wording so minor
   orchestrator phrasing changes keep matching, but it requires the
   "verify provider" core, so a rejected key, a missing runtime, or a crashed
   command never triggers the fallback. A response that already carries a
   connected provider never matches.
2. **Verify the key against the provider API.** Anthropic is the one provider
   with an app-side verifier today (`GET /v1/models`). The verifier moves out of
   `ProviderConnectionService` into `src/main/utils/anthropicApiKeyVerification.ts`
   so the connection service and this fallback share one implementation instead
   of drifting; `ProviderConnectionService` loses ~90 lines and gains an import.
3. **Commit, re-read, roll back.** Only on a `valid` verdict: atomically merge
   the credential into the runtime's `opencode/auth.json` (bounded read, refuses
   a non-JSON or oversized store rather than clobbering other providers' logins),
   invalidate the provider caches, and re-run the provider view. If the provider
   still does not report `connected`, the entry is rolled back to its previous
   value — and if something else replaced it in the meantime, it is left alone.

This keeps every guarantee of the verify-on-connect behaviour added for #276: an
`invalid` verdict, an inconclusive one, a non-probe failure, or a provider with
no app-side verifier all leave the response exactly as the runtime returned it,
and nothing is ever written to the auth store in those cases. The only change is
that a key the probe *cannot* verify — as opposed to one it verified as bad — is
no longer refused.

The second commit routes the setup-form path through the same recovery.
`resolveConnectRecoveryInput` normalizes either connect input shape onto the
recovery input and returns `null` for OAuth connects and for connects with no
key, so those paths are untouched.

### Note on the CLI client refactor

`AgentTeamsRuntimeProviderManagementCliClient.ts` is a frozen legacy file whose
size cap may only move downward, so the wiring had to pay for itself.
`connectProvider` and `connectWithApiKey` carried an identical 23-line response
tail (parse JSON on success → sanitized error payload → raw non-JSON output as a
command failure). That tail is now one `parseProviderCommandResponse` helper,
which is also the single point where each path applies the recovery. The file
ends at its existing cap of 2685 lines with no baseline change.

## Tests

- `test/main/features/runtime-provider-management/openCodeConnectVerifyFailure.test.ts`
  (new, 5 tests) — matches the live orchestrator message, rephrased and
  contracted wording, the signature across line breaks and inside ANSI-coded
  diagnostics previews; ignores unrelated auth failures and non-error responses;
  never matches a response that already carries a connected provider.
- `test/main/features/runtime-provider-management/openCodeConnectApiKeyFallback.test.ts`
  (new, 10 tests) — the happy path (commit, re-read, connected provider); no
  write when the API rejects the key; no write when verification is
  inconclusive; rollback when the provider still is not connected; rollback that
  restores an older entry; rollback when the status re-read itself throws; no
  activation without an app-side verifier; no activation without the probe
  signature; abort without writing when the auth store is not valid JSON; a
  concurrently replaced credential is left alone instead of being rolled back.
- `test/main/features/runtime-provider-management/AgentTeamsRuntimeProviderManagementCliClient.test.ts`
  (+3, 76 total) — end-to-end through the client: `connectWithApiKey` commits the
  verified key and re-reads status; `connectProvider` (setup form) recovers the
  same way; a non-probe failure is returned unchanged with no verification call
  and no auth-store file created. These stub the data home to a temp dir so the
  auth-store write never touches a real profile.

All green: 6 files / 201 tests (including the existing
`test/main/services/runtime/ProviderConnectionService.test.ts`, 75 tests, which
pins that moving the verifier changed no behaviour).

## Tested on

Windows 11 (24H2), Electron build from source, packaged orchestrator runtime.

Provider combinations actually exercised for this change:

- **Anthropic API key through OpenCode** — both entry points: the provider
  connect dialog (`connect-api-key`) and the provider setup form
  (`runtime providers connect --method api`). Reproduced the original
  `Not Found` candidate failure with a known-good key, confirmed the key is
  committed to `opencode/auth.json` and the provider flips to connected, and
  confirmed a deliberately wrong key still fails with the provider's own
  rejection and writes nothing.
- **Local models via OpenCode** (llama-swap OpenAI-compatible endpoint) —
  confirmed unaffected: no app-side verifier is registered, so those connects
  take the unchanged path.
- **Anthropic OAuth through OpenCode** — confirmed the OAuth connect path is
  skipped by `resolveConnectRecoveryInput` and behaves exactly as before.

Not applicable to this change: Codex (ChatGPT or API key) and Cursor, which do
not connect through the OpenCode provider dialogs. Only Anthropic has an
app-side verifier today; other OpenCode providers keep the current behaviour
until one is added for them. Not tested on macOS or Linux — the auth-store path
resolution is shared platform code (`resolveClaudeMultimodelDataHomePath`), but
the fix was only exercised on Windows.

## References

Refs #276 — completes the verify-on-connect behaviour added there. It keeps that
behaviour's guarantees for real credential failures (a key the provider rejects
is still refused, and nothing is written), and stops it from refusing keys the
runtime-side probe merely *cannot* verify because the probe runs before the key
is active.

## Validation checklist

- [x] `pnpm typecheck` — exit 0
- [x] Targeted `vitest run` on the files above — 201/201 passing
- [x] `pnpm guard:source-file-size` — passing, no baseline change; the CLI client
      stays at its existing 2685-line cap
- [x] `pnpm lint` on the changed source files — 0 errors
- [x] Branched from and rebased on current `dev`

---
**Authorship & provenance:** All code in this PR was written by **Claude (Fable 5)**, directed and live-verified on Windows by me. As with anything under AGPL-3.0, it is offered **as-is, without warranty of any kind** - use, adapt, or reject freely. Full context: Discord thread, Aug 30. Review questions are welcome; answers will come from the same Claude-assisted workflow.

## Summary by CodeRabbit

* **New Features**
  * Added recovery for OpenCode provider connection verification failures.
  * API keys can now be verified directly and saved when runtime verification requires prior activation.
  * Added secure Anthropic API-key verification with timeout and redirect protection.
* **Bug Fixes**
  * Prevented invalid credentials from being saved and restored provider state when recovery fails.
  * Improved provider status refresh after connection recovery attempts.
* **Tests**
  * Added coverage for successful recovery, rejected keys, rollback behavior, cache refreshes, and verification edge cases.

### CODE REVIEW

File: src/main/utils/anthropicApiKeyVerification.ts [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+  baseUrl?: string | null
+): Promise<AnthropicApiKeyVerificationResult> {
+  const controller = new AbortController();
+  const timeout = setTimeout(() => controller.abort(), ANTHROPIC_API_KEY_VERIFY_TIMEOUT_MS);
+  try {
+    const response = await fetch(buildAnthropicModelsUrl(baseUrl), {
+      method: 'GET',
+      signal: controller.signal,
+      headers: {
+        'x-api-key': apiKey,
```
[the remark this answers is not in the document]
Comment (NeatCode-Labs) [reply src/main/utils/anthropicApiKeyVerification.ts:66]: Good catch — fixed in 5b4220306. I confirmed the undici behavior directly rather than from docs: on Node 24, a cross-origin redirect strips `Authorization` but retains `x-api-key`, so the probe could have forwarded the key to an unconfigured origin. Added `redirect: 'error'`; the resulting `TypeError` lands in the existing `unknown` branch, so a refused redirect is reported as "could not verify" and never as an invalid key, leaving the connect fallback's recovery intact. One note: this file is an extraction of upstream's inline verifier, which had the identical default-redirect fetch (introduced in 6dc103b73) — pre-existing upstream behavior, hardened here.

File: src/features/runtime-provider-management/main/infrastructure/openCodeConnectVerifyFailure.ts [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+ * catalog model candidates, e.g. "OpenCode could not verify provider anthropic
+ * with 3 model candidates: ...: Not Found". OpenCode only activates a provider
+ * once its credential is in the auth store, so for store-activated providers
+ * that probe runs without the submitted key and every candidate fails. The
+ * matcher stays deliberately loose about wording so minor orchestrator
+ * phrasing changes keep matching, while still requiring the "verify provider"
+ * core so unrelated auth failures never trigger the fallback.
+ */
+const VERIFY_FAILURE_PATTERN =
+  /\b(?:could ?n[o']t|cannot|can ?not|can't|unable to|failed to) verify (?:the )?provider\b/i;
```
[the remark this answers is not in the document]
Comment (NeatCode-Labs) [reply src/features/runtime-provider-management/main/infrastructure/openCodeConnectVerifyFailure.ts:21]: You're right that the pattern matches that string — the orchestrator folds per-candidate reasons into the same "could not verify provider … model candidates" message, so an auth reason can appear there. It could not persist a rejected key, though: the fallback commits only after an independent `GET /v1/models` against the provider API returns 200, and 401/403 map to `invalid`, which returns the original response untouched. I've narrowed the classifier anyway so it never admits such a failure: any rejection evidence in the message or diagnostics (invalid/expired API key, invalid credentials, `authentication_error`, `not_authenticated`, unauthorized/forbidden, bare 401/403) now vetoes the match, with tests in both directions and the bare-status match kept clear of model-id digits.

File: src/features/runtime-provider-management/main/infrastructure/openCodeConnectApiKeyFallback.ts [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+  providerId: string;
+  apiKey: string;
+}): Promise<OpenCodeAuthStoreCredentialSnapshot> {
+  const store = await readOpenCodeAuthStore(input.authStorePath);
+  const snapshot: OpenCodeAuthStoreCredentialSnapshot = {
+    previousCredential: store[input.providerId],
+    hadPreviousCredential: Object.hasOwn(store, input.providerId),
+  };
+  store[input.providerId] = { type: 'api', key: input.apiKey };
+  await writeOpenCodeAuthStore(input.authStorePath, store);
```
[the remark this answers is not in the document]
Comment (NeatCode-Labs) [reply src/features/runtime-provider-management/main/infrastructure/openCodeConnectApiKeyFallback.ts]: Real, though narrower than described — both paths already re-read fresh and merge only our provider key, and the write is a temp+fsync+rename, so no torn file and no stale-snapshot clobber of other providers. The remaining window was read→rename, which a concurrent writer could still lose, so both writes now go through one helper that uses `atomicWriteAsync`'s `beforeCommit` seam to re-read and refuse to publish over changed content, retrying against the freshest snapshot. That is not a true compare-and-swap — the store layer has no file locking — but it shrinks the window from the whole stage-and-fsync to one rename; a concurrent-writer test covers it.

File: src/features/runtime-provider-management/main/infrastructure/openCodeConnectApiKeyFallback.ts [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+      response,
+      `The app then verified the API key directly with the ${providerId} API and committed it to the OpenCode auth store, but OpenCode still did not report the provider as connected, so ${rollbackNote}`,
+      [
+        'The API key itself is valid: the app verified it directly against the provider API.',
+        'OpenCode did not pick up the committed credential; refresh the provider catalog and try again.',
+      ]
+    );
+  } catch {
+    // The fallback must never make the original failure worse.
+    return response;
```
[the remark this answers is not in the document]
Comment (NeatCode-Labs) [reply src/features/runtime-provider-management/main/infrastructure/openCodeConnectApiKeyFallback.ts:336]: Confirmed and fixed. Only the "provider is not connected" branch rolled back; a host that threw after the commit — `loadView` rejecting, or cache invalidation throwing — fell through to the outer catch and returned the original failure with the new key still in the auth store, which contradicts the documented contract. The commit is now tracked and undone from that catch as well, so every path that reports failure leaves the store as it found it, covered by tests for both throw sites.
